### PR TITLE
Gm focus

### PR DESCRIPTION
**Pull request by mgalbier**
- Improve focus handling:
  - Have new attribute in NodeReviews to store focusInformation
  - Move visualization of header to VNodeReviews
  - Added keyboard shortcut ctrl+F12 to perform focus step
  - Changed syntax for focusInformation
  - Allow multiple FocusInformation entries for the same step
  - Allow FocusTypes to be combined
- Move functionality to center and zoom view from CSceneHandlerLoad to MainView
- Use functionality after loading Diff, History and Review
  <a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72472939%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72473106%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72473365%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72473529%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72473808%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72474024%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72474036%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72474398%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72474539%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72474657%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72475197%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23issuecomment-235644634%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72580836%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72580845%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72581204%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72581322%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72581569%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23issuecomment-235835926%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23issuecomment-235644634%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Alright%2C%20I%27m%20done%20with%20the%20review.%20Mostly%20smaller%20tweaks%20are%20needed.%22%2C%20%22created_at%22%3A%20%222016-07-27T16%3A39%3A16Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Thanks.%22%2C%20%22created_at%22%3A%20%222016-07-28T08%3A45%3A21Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20e67b8a190f987714f943fbaec748dca426921359%20VisualizationBase/src/views/MainView.h%204%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72475197%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Please%20rename%20this%20to%20%60centerAndZoomViewToFitEntireScene%60%22%2C%20%22created_at%22%3A%20%222016-07-27T16%3A38%3A16Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VisualizationBase/src/views/MainView.h%3AL49-57%22%7D%2C%20%22Pull%20f9cd2f26079e5fb6a60f75a60348266301469a79%20CodeReview/src/commands/CFocus.cpp%20146%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72474539%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22The%20best%20option%20is%20to%20add%20%60isValid%60%20method%20to%20%60FocusInformation%60%20and%20define%20there%20what%20that%20means%2C%20not%20here.%22%2C%20%22created_at%22%3A%20%222016-07-27T16%3A34%3A35Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20CodeReview/src/commands/CFocus.cpp%3AL113-158%22%7D%2C%20%22Pull%20f9cd2f26079e5fb6a60f75a60348266301469a79%20CodeReview/src/commands/CFocus.cpp%20125%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72474398%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Perhaps%20before%20returning%20this%2C%20you%20should%20make%20sure%20that%20it%20is%20valid%3F%20This%20is%20just%20to%20future%20proof%20the%20system.%20If%20you%20make%20the%20FocusInfo%20have%20%60None%60%20as%20type%20by%20default%2C%20and%20%60-1%60%20as%20the%20step.%20Then%20a%20valid%20focus%20info%20is%20one%20which%20has%20both%20of%20these%20different.%22%2C%20%22created_at%22%3A%20%222016-07-27T16%3A33%3A44Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20CodeReview/src/commands/CFocus.cpp%3AL113-158%22%7D%2C%20%22Pull%20f9cd2f26079e5fb6a60f75a60348266301469a79%20CodeReview/src/items/VNodeReviewsStyle.h%2013%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72472939%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22This%20property%20seems%20to%20be%20unused%20%28also%20before%20the%20current%20change%29.%20Delete%20it.%22%2C%20%22created_at%22%3A%20%222016-07-27T16%3A24%3A49Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20CodeReview/src/items/VNodeReviewsStyle.h%3AL41-51%22%7D%2C%20%22Pull%20f9cd2f26079e5fb6a60f75a60348266301469a79%20CodeReview/src/commands/CFocus.h%2023%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72474657%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22The%20step%20above%2C%20should%20probably%20be%20%60-1%60%20by%20default%2C%20also%20indicating%20an%20invalid%20info.%22%2C%20%22created_at%22%3A%20%222016-07-27T16%3A35%3A14Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20CodeReview/src/commands/CFocus.h%3AL39-55%22%7D%2C%20%22Pull%20f9cd2f26079e5fb6a60f75a60348266301469a79%20CodeReview/src/commands/CFocus.cpp%2095%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72474024%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Use%20the%20better%20condition%3A%20%60if%20%28QString%7B%5C%22highlight%5C%22%7D.startsWith%28token%29%29%60%22%2C%20%22created_at%22%3A%20%222016-07-27T16%3A31%3A31Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20CodeReview/src/commands/CFocus.cpp%3AL113-158%22%7D%2C%20%22Pull%20f9cd2f26079e5fb6a60f75a60348266301469a79%20CodeReview/src/commands/CFocus.cpp%2053%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72473808%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Hmm%2C%20somewhere%2C%20we%20should%20remove%20the%20highlights%20of%20the%20previous%20step.%20Please%20add%20that%20code%20and%20test%20it.%22%2C%20%22created_at%22%3A%20%222016-07-27T16%3A30%3A10Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22At%20the%20beginning%20of%20the%20method%20I%20call%20%60removeOverlayGroup%28%5C%22focusOverlay%5C%22%29%60.%20This%20seems%20to%20work.%22%2C%20%22created_at%22%3A%20%222016-07-28T08%3A14%3A17Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/17177211%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mgalbier%22%7D%7D%2C%20%7B%22body%22%3A%20%22Yes%20indeed.%20I%20missed%20that%20%3A%29%22%2C%20%22created_at%22%3A%20%222016-07-28T08%3A17%3A12Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20CodeReview/src/commands/CFocus.cpp%3AL63-107%22%7D%2C%20%22Pull%20f9cd2f26079e5fb6a60f75a60348266301469a79%20CodeReview/src/commands/CFocus.cpp%2018%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72473365%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Instead%20of%20using%20the%20%60mainScene%60%20use%20the%20scene%20of%20the%20%60target%60%20argument%2C%20which%20you%27re%20currently%20not%20using.%22%2C%20%22created_at%22%3A%20%222016-07-27T16%3A27%3A23Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22That%20should%20also%20make%20it%20unnecessary%20to%20%60%23include%20%5C%22VisualizationBase/src/VisualizationManager.h%5C%22%60%22%2C%20%22created_at%22%3A%20%222016-07-27T16%3A28%3A24Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20also%20use%20%60Visualization%3A%3AVisualizationManager%3A%3Ainstance%28%29.mainView%28%29-%3EcenterOn%28focusItem%29%3B%60.%20Is%20there%20also%20a%20way%20to%20do%20this%20using%20target%3F%22%2C%20%22created_at%22%3A%20%222016-07-28T08%3A14%3A11Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/17177211%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mgalbier%22%7D%7D%2C%20%7B%22body%22%3A%20%22Hmm%2C%20for%20that%20you%20should%20better%20keep%20using%20your%20current%20approach%20%28and%20keep%20the%20include%29%2C%20but%20still%20use%20the%20target%27s%20scene%20for%20the%20line%20below.%22%2C%20%22created_at%22%3A%20%222016-07-28T08%3A18%3A17Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Ok%2C%20will%20do.%22%2C%20%22created_at%22%3A%20%222016-07-28T08%3A19%3A49Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/17177211%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mgalbier%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20CodeReview/src/commands/CFocus.cpp%3AL63-107%22%7D%2C%20%22Pull%20f9cd2f26079e5fb6a60f75a60348266301469a79%20CodeReview/src/commands/CFocus.cpp%2097%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72474036%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Like%20above.%22%2C%20%22created_at%22%3A%20%222016-07-27T16%3A31%3A38Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20CodeReview/src/commands/CFocus.cpp%3AL113-158%22%7D%2C%20%22Pull%20f9cd2f26079e5fb6a60f75a60348266301469a79%20CodeReview/styles/item/VNodeReviews/default%204%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/438%23discussion_r72473106%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22remove%22%2C%20%22created_at%22%3A%20%222016-07-27T16%3A25%3A51Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20CodeReview/styles/item/VNodeReviews/default%3AL16-30%22%7D%7D%7D'></a>
  <a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-Pull f9cd2f26079e5fb6a60f75a60348266301469a79 CodeReview/src/items/VNodeReviewsStyle.h 13'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/438#discussion_r72472939'>File: CodeReview/src/items/VNodeReviewsStyle.h:L41-51</a></b>
- <a href='https://github.com/dimitar-asenov'></a> This property seems to be unused (also before the current change). Delete it.
- [ ] <a href='#crh-comment-Pull f9cd2f26079e5fb6a60f75a60348266301469a79 CodeReview/styles/item/VNodeReviews/default 4'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/438#discussion_r72473106'>File: CodeReview/styles/item/VNodeReviews/default:L16-30</a></b>
- <a href='https://github.com/dimitar-asenov'></a> remove
- [ ] <a href='#crh-comment-Pull f9cd2f26079e5fb6a60f75a60348266301469a79 CodeReview/src/commands/CFocus.cpp 18'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/438#discussion_r72473365'>File: CodeReview/src/commands/CFocus.cpp:L63-107</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Instead of using the `mainScene` use the scene of the `target` argument, which you're currently not using.
- <a href='https://github.com/dimitar-asenov'></a> That should also make it unnecessary to `#include "VisualizationBase/src/VisualizationManager.h"`
- <a href='https://github.com/mgalbier'></a> I also use `Visualization::VisualizationManager::instance().mainView()->centerOn(focusItem);`. Is there also a way to do this using target?
- <a href='https://github.com/dimitar-asenov'></a> Hmm, for that you should better keep using your current approach (and keep the include), but still use the target's scene for the line below.
- <a href='https://github.com/mgalbier'></a> Ok, will do.
- [ ] <a href='#crh-comment-Pull f9cd2f26079e5fb6a60f75a60348266301469a79 CodeReview/src/commands/CFocus.cpp 53'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/438#discussion_r72473808'>File: CodeReview/src/commands/CFocus.cpp:L63-107</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Hmm, somewhere, we should remove the highlights of the previous step. Please add that code and test it.
- <a href='https://github.com/mgalbier'></a> At the beginning of the method I call `removeOverlayGroup("focusOverlay")`. This seems to work.
- <a href='https://github.com/dimitar-asenov'></a> Yes indeed. I missed that :)
- [ ] <a href='#crh-comment-Pull f9cd2f26079e5fb6a60f75a60348266301469a79 CodeReview/src/commands/CFocus.cpp 95'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/438#discussion_r72474024'>File: CodeReview/src/commands/CFocus.cpp:L113-158</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Use the better condition: `if (QString{"highlight"}.startsWith(token))`
- [ ] <a href='#crh-comment-Pull f9cd2f26079e5fb6a60f75a60348266301469a79 CodeReview/src/commands/CFocus.cpp 97'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/438#discussion_r72474036'>File: CodeReview/src/commands/CFocus.cpp:L113-158</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Like above.
- [ ] <a href='#crh-comment-Pull f9cd2f26079e5fb6a60f75a60348266301469a79 CodeReview/src/commands/CFocus.cpp 125'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/438#discussion_r72474398'>File: CodeReview/src/commands/CFocus.cpp:L113-158</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Perhaps before returning this, you should make sure that it is valid? This is just to future proof the system. If you make the FocusInfo have `None` as type by default, and `-1` as the step. Then a valid focus info is one which has both of these different.
- [ ] <a href='#crh-comment-Pull f9cd2f26079e5fb6a60f75a60348266301469a79 CodeReview/src/commands/CFocus.cpp 146'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/438#discussion_r72474539'>File: CodeReview/src/commands/CFocus.cpp:L113-158</a></b>
- <a href='https://github.com/dimitar-asenov'></a> The best option is to add `isValid` method to `FocusInformation` and define there what that means, not here.
- [ ] <a href='#crh-comment-Pull f9cd2f26079e5fb6a60f75a60348266301469a79 CodeReview/src/commands/CFocus.h 23'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/438#discussion_r72474657'>File: CodeReview/src/commands/CFocus.h:L39-55</a></b>
- <a href='https://github.com/dimitar-asenov'></a> The step above, should probably be `-1` by default, also indicating an invalid info.
- [ ] <a href='#crh-comment-Pull e67b8a190f987714f943fbaec748dca426921359 VisualizationBase/src/views/MainView.h 4'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/438#discussion_r72475197'>File: VisualizationBase/src/views/MainView.h:L49-57</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Please rename this to `centerAndZoomViewToFitEntireScene`
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/438#issuecomment-235644634'>General Comment</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Alright, I'm done with the review. Mostly smaller tweaks are needed.
- <a href='https://github.com/dimitar-asenov'></a> Thanks.

<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/438?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/438?approve=1'></a>&nbsp;<a href='https://github.com/dimitar-asenov/Envision/pull/438'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by dimitar-asenov** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/dimitar-asenov/Envision/pull/438'>Refresh</a>.<a href='#crh-update-none'></a>

File: CodeReview/src/items/VNodeReviewsStyle.h [hunk 1]
```diff
@@ -39,6 +41,10 @@ class CODEREVIEW_API VNodeReviewsStyle : public Super<Visualization::Declarative
 	public:
 		virtual ~VNodeReviewsStyle() override;
 		Property<Visualization::VListStyle> comments{this, "comments"};
+		Property<Visualization::TextStyle> commentInput{this, "commentInput"};
```
Comment (dimitar-asenov) [inline CodeReview/src/items/VNodeReviewsStyle.h]: This property seems to be unused (also before the current change). Delete it.

File: CodeReview/styles/item/VNodeReviews/default [hunk 2]
```diff
@@ -16,4 +16,14 @@
 		</shape>
 		<useBackgroundColors>false</useBackgroundColors>
 	</comments>
+	<commentInput prototypes="item/Text/default" />
```
Comment (dimitar-asenov) [inline CodeReview/styles/item/VNodeReviews/default]: remove

File: CodeReview/src/commands/CFocus.cpp [hunk 3]
```diff
@@ -61,41 +63,44 @@ bool CFocus::canInterpret(Visualization::Item*, Visualization::Item*,
 Interaction::CommandResult* CFocus::execute(Visualization::Item*, Visualization::Item*,
 				const QStringList&, const std::unique_ptr<Visualization::Cursor>&)
 {
+	focusStep({}, {}, {});
+	return new Interaction::CommandResult{};
+}
 
+bool CFocus::focusStep(Visualization::Item *, QKeySequence, Interaction::ActionRegistry::InputState)
+{
```
Comment (dimitar-asenov) [inline CodeReview/src/commands/CFocus.cpp:71]: Instead of using the `mainScene` use the scene of the `target` argument, which you're currently not using.

Comment (dimitar-asenov) [inline CodeReview/src/commands/CFocus.cpp:71]: That should also make it unnecessary to `#include "VisualizationBase/src/VisualizationManager.h"`

File: CodeReview/src/commands/CFocus.cpp [hunk 4]
```diff
@@ -61,41 +63,44 @@ bool CFocus::canInterpret(Visualization::Item*, Visualization::Item*,
...
-			Visualization::VisualizationManager::instance().mainView()->
-					centerOn(focusItem);
-			break;
-		case FocusInformation::Highlight:
+		if (focusInformation.type_.testFlag(FocusInformation::Center))
+		{
+			Visualization::VisualizationManager::instance().mainView()->centerOn(focusItem);
+		}
+
+		if (focusInformation.type_.testFlag(FocusInformation::Highlight))
```
Comment (dimitar-asenov) [inline CodeReview/src/commands/CFocus.cpp:94]: Hmm, somewhere, we should remove the highlights of the previous step. Please add that code and test it.

File: CodeReview/src/commands/CFocus.cpp [hunk 5]
```diff
@@ -108,59 +113,45 @@ QList<Interaction::CommandSuggestion*> CFocus::suggest(Visualization::Item*, Vis
...
-		commandTokens.pop_front();
+	FocusInformation focusInformation;
 
-		FocusInformation focusInformation;
+	while (!commandTokens.isEmpty())
+	{
+		auto token = commandTokens.takeFirst();
 
-		if (!commandTokens.isEmpty())
+		if (token.startsWith("h"))
```
Comment (dimitar-asenov) [inline CodeReview/src/commands/CFocus.cpp]: Use the better condition: `if (QString{"highlight"}.startsWith(token))`

File: CodeReview/src/commands/CFocus.cpp [hunk 6]
```diff
@@ -108,59 +113,45 @@ QList<Interaction::CommandSuggestion*> CFocus::suggest(Visualization::Item*, Vis
...
+			focusInformation.type_ |= FocusInformation::FocusType::Highlight;
+		else if (token.startsWith("c"))
```
Comment (dimitar-asenov) [inline CodeReview/src/commands/CFocus.cpp]: Like above.

File: CodeReview/src/commands/CFocus.cpp [hunk 7]
```diff
@@ -108,59 +113,45 @@ QList<Interaction::CommandSuggestion*> CFocus::suggest(Visualization::Item*, Vis
...
-		if (!commandTokens.isEmpty())
-		{
-			auto stepToken = commandTokens.takeFirst();
-			focusInformation.step_ = stepToken.toInt();
-		}
-		return focusInformation;
 	}
 
-	return {};
+	return focusInformation;
```
Comment (dimitar-asenov) [inline CodeReview/src/commands/CFocus.cpp:142]: Perhaps before returning this, you should make sure that it is valid? This is just to future proof the system. If you make the FocusInfo have `None` as type by default, and `-1` as the step. Then a valid focus info is one which has both of these different.

File: CodeReview/src/commands/CFocus.cpp [hunk 8]
```diff
@@ -108,59 +113,45 @@ QList<Interaction::CommandSuggestion*> CFocus::suggest(Visualization::Item*, Vis
...
-		if (firstComment->lines()->isEmpty())
-			continue;
-
-		auto firstLine = firstComment->lines()->first()->get();
-
-		auto focusInformation = extractFocusInformation(firstLine);
+		auto focusInformation = extractFocusInformation(nodeReviews->focusInformation());
 
-		if (focusInformation.type_ != FocusInformation::None)
+		if (!focusInformation.type_.testFlag(FocusInformation::None))
```
Comment (dimitar-asenov) [inline CodeReview/src/commands/CFocus.cpp]: The best option is to add `isValid` method to `FocusInformation` and define there what that means, not here.

File: CodeReview/src/commands/CFocus.h [hunk 9]
```diff
@@ -38,10 +39,16 @@ class CODEREVIEW_API CFocus : public Interaction::Command
...
+				None = 0x0,
+				Center = 0x1,
+				Highlight= 0x2
+			};
+
+			using FocusTypes = QFlags<FocusType>;
 
 			int step_{};
-			FocusType type_{};
+			FocusTypes type_{};
```
Comment (dimitar-asenov) [inline CodeReview/src/commands/CFocus.h:51]: The step above, should probably be `-1` by default, also indicating an invalid info.

File: VisualizationBase/src/views/MainView.h [hunk 10]
```diff
@@ -49,6 +49,8 @@ class VISUALIZATIONBASE_API MainView: public View
 		static const int MINIMAP_DEFAULT_HEIGHT = 200;
 		static const int PNG_SCREENSHOT_SCALE = 8;
 
+		static void centerAndZoomView();
```
Comment (dimitar-asenov) [inline VisualizationBase/src/views/MainView.h]: Please rename this to `centerAndZoomViewToFitEntireScene`

**Comment by dimitar-asenov** [comment]
Alright, I'm done with the review. Mostly smaller tweaks are needed.

File: CodeReview/src/commands/CFocus.cpp (hunk 3 above)
Comment (mgalbier) [inline CodeReview/src/commands/CFocus.cpp:71]: I also use `Visualization::VisualizationManager::instance().mainView()->centerOn(focusItem);`. Is there also a way to do this using target?

File: CodeReview/src/commands/CFocus.cpp (hunk 4 above)
Comment (mgalbier) [inline CodeReview/src/commands/CFocus.cpp:94]: At the beginning of the method I call `removeOverlayGroup("focusOverlay")`. This seems to work.

Comment (dimitar-asenov) [inline CodeReview/src/commands/CFocus.cpp:94]: Yes indeed. I missed that :)

File: CodeReview/src/commands/CFocus.cpp (hunk 3 above)
Comment (dimitar-asenov) [inline CodeReview/src/commands/CFocus.cpp:71]: Hmm, for that you should better keep using your current approach (and keep the include), but still use the target's scene for the line below.

Comment (mgalbier) [inline CodeReview/src/commands/CFocus.cpp:71]: Ok, will do.

**Comment by dimitar-asenov** [comment]
Thanks.

**Pull request merged**